urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/updater/FileExtractor.java

```diff
@@ -0,0 +1,218 @@
+        
+        return uncompressed;
+    }
+    
+    @SuppressWarnings("unchecked")
+    public static void extractZIP(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        final ZipFile zip = new ZipFile(file);
+        
+        Enumeration<ZipEntry> entries = (Enumeration<ZipEntry>) zip.entries();
+        
+        int total = 0;
+        while (entries.hasMoreElements())
+        {
+            final ZipEntry entry = entries.nextElement();
+            
+            if (!entry.isDirectory())
+            {
+                total += entry.getSize();
+            }
+        }
+        
+        entries = (Enumeration<ZipEntry>) zip.entries();
+        final int current = 0;
+        
+        while (entries.hasMoreElements())
+        {
+            final ZipEntry entry = entries.nextElement();
+            
+            if (entry.isDirectory())
+            {
+                final File dir = new File(dest, entry.getName());
+                if (!dir.exists())
+                {
+                    dir.mkdirs();
+                }
+            }
+            else
+            {
+                final File e = new File(dest, entry.getName());
+                copyStream(api, source, zip.getInputStream(entry),
+                        new FileOutputStream(e), current, total, min, max);
+            }
+        }
+        
+        zip.close();
+    }
+    
+    public static void extractJAR(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        JarFile jar = null;
+        try
+        {
+            jar = new JarFile(file);
+        }
+        catch (final IOException e)
+        {
+            e.printStackTrace();
+        }
+        
+        Enumeration<JarEntry> entries = jar.entries();
+        
+        int total = 0;
+        while (entries.hasMoreElements())
+        {
+            final JarEntry entry = entries.nextElement();
+            
+            if (!entry.isDirectory()
+                    && !entry.getName().equalsIgnoreCase("MANIFEST.MF"))
+            {
+                total += entry.getSize();
+            }
+        }
+        
+        entries = jar.entries();
+        final int current = 0;
+        
+        while (entries.hasMoreElements())
+        {
+            final JarEntry entry = entries.nextElement();
+            
+            if (entry.isDirectory())
+            {
+                final File dir = new File(dest, entry.getName());
+                if (!dir.exists()
+                        && !dir.getName().equalsIgnoreCase("META-INF"))
+                {
+                    dir.mkdirs();
+                }
+            }
+            else
+            {
+                final File e = new File(dest, entry.getName());
+                if (!e.getParent().contains("META-INF"))
+                {
+                    copyStream(api, source, jar.getInputStream(entry),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource returned by the following method call: *getInputStream*. The resource is referenced at lines 140, 150, 164, 185. The resource is closed at line 215. There are other execution paths that do not contain closure statements, for example, when InputStream.read (in method copyStream) throws an exception. Ensure the object returned by *getInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).